modyn/supervisor/internal/triggers/performance/decision_policy.py

```diff
@@ -0,0 +1,202 @@
+    ) -> bool:
+        """Utilizes the state of `DataDensityTracker` and `PerformanceTracker`
+        to make the decision.
+
+        We support both the "hindsight" and "forecast" mode.
+
+        In the "hindsight" mode, the decision is made based on the current performance and the cumulated avoidable misclassifications.
+
+        - Formalization:
+            - historic observation:
+                - data_cum_since_last_trigger: The cumulated data points since the last trigger.
+                - avoidable_misclassifications_since_last_trigger: The cumulated avoidable misclassifications since
+                    the last trigger.
+
+        In the "lookahead" mode, the decision is made based on the current performance, the cumulated avoidable
+        misclassifications, future performance estimates and future data density estimates.
+        Similar to the "hindsight" mode, we first check if current performance already leads to a transgression
+        of the threshold and therefore to a trigger.
+
+        If that's not the case we estimate the cumulated avoidable misclassifications until the next point of update.
+        If we expect a transgression of the threshold before the next update point, we trigger.
+
+        This forward looking approach tries to avoid exceeding the misclassification budget in the first place.
+        """
+
+        # compute the number of avoidable misclassifications by retrieving the actual misclassifications
+        # and the expected misclassifications through the expected accuracy for the last interval.
+        previous_interval_num_misclassifications = performance_tracker.previous_batch_num_misclassifications()
+
+        # the expected performance won't change unless there's a trigger
+        expected_accuracy = (
+            performance_tracker.forecast_expected_accuracy(method=method)
+            if self.config.expected_accuracy is None
+            else self.config.expected_accuracy
+        )
+        previous_interval_num_samples = data_density.previous_batch_samples()
+        previous_expected_num_misclassifications = (1 - expected_accuracy) * previous_interval_num_samples
+        new_avoidable_misclassifications = (
+            previous_interval_num_misclassifications - previous_expected_num_misclassifications
+        )
+        if new_avoidable_misclassifications < 0 and not self.config.allow_reduction:
+            new_avoidable_misclassifications = 0
+
+        self.cumulated_avoidable_misclassifications += round(new_avoidable_misclassifications)
+
+        if mode == "hindsight":
+            return self.cumulated_avoidable_misclassifications >= self.config.avoidable_misclassification_threshold
+
+        if mode == "lookahead":
+            # past misclassifications already exceed the threshold, forecasting not needed
+            if self.cumulated_avoidable_misclassifications >= self.config.avoidable_misclassification_threshold:
+                return True
```

Comment:
What I don't see yet is how we reset this after a trigger. Otherwise we will always keep triggering, won't we? (The threshold might increase but my intuition would be to reset the cumulation buffer after trigger)

Reply:
```python
    def inform_trigger(self) -> None:
        """Resets the cumulated avoidable misclassifications."""
        self.cumulated_avoidable_misclassifications = 0
```